common/src/main/java/org/opensearch/ml/common/connector/HttpConnector.java

```diff
@@ -264,6 +265,64 @@ public  <T> T createPredictPayload(Map<String, String> parameters) {
         return (T) parameters.get("http_body");
     }
 
+    protected String applyCustomSubstitution(Map<String, String> parameters, String inputPayload) {
+
+        if (parameters == null) {
+            throw new IllegalArgumentException("Some parameter placeholder not filled in payload: input");
+        }
+        String result = inputPayload;
+
+        Pattern pattern = Pattern.compile("\\$\\{parameters\\.(\\w+)\\}");
+        Matcher matcher = pattern.matcher(inputPayload);
+
+        while (matcher.find()) {
+            String paramName = matcher.group(1);
+            String paramValue = parameters.get(paramName);
+            if (paramValue != null) {
+                String modifiedValue = isNotPlainString(paramValue) ? paramValue : "\"" + paramValue + "\"";
+                result = result.replace("${parameters." + paramName + "}", modifiedValue);
+            }
+        }
+
+        result = StringEscapeUtils.unescapeJava(result); // Unescape Unicode escape sequences
```

Comment:
Here, I am not sure if this is sufficient to avoid valid values causing `isJson()` to fail as I mentioned above.

Reply:
Could you please explain this little bit more for me?

Reply:
Yaliang answered my question above.